src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,294 @@
+                            .replace ("</ol>", "\n")
+                            .strip ();
+                            break;
+                        case "Protocol":
+                            release.protocol = val.get_string ();
+                            break;
+                        case "RemoteId":
+                            release.remote_id = val.get_string ();
+                            break;
+                        case "AppstreamId":
+                            release.appstream_id = val.get_string ();
+                            break;
+                        case "Checksum":
+                            release.checksum = val.get_string ();
+                            break;
+                        case "Vendor":
+                            release.vendor = val.get_string ();
+                            break;
+                        case "Size":
+                            release.size = val.get_uint64 ();
+                            break;
+                        case "License":
+                            release.license = val.get_string ();
+                            break;
+                        case "TrustFlags":
+                            release.flag = ReleaseFlag.from_uint64 (val.get_uint64 ());
+                            break;
+                        case "InstallDuration":
+                            release.install_duration = val.get_uint32 ();
+                            break;
+                        case "Uri":
+                            release.uri = val.get_string ();
+                            break;
+                        default:
+                            break;
+                    }
+                }
+                releases_list.append (release);
+            }
+        } catch (Error e) {
+            warning ("Could not connect to fwupd interface: %s", e.message);
+        }
+
+        return releases_list;
+    }
+
+    private string get_path (Release release) {
+        var parts = release.uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async void install (string id, Release release) {
+        var path = get_path (release);
+
+        File server_file = File.new_for_uri (release.uri);
+        File local_file = File.new_for_path (path);
+
+        bool result;
+        try {
+            result = yield server_file.copy_async (local_file, FileCopyFlags.OVERWRITE, Priority.DEFAULT, null, (current_num_bytes, total_num_bytes) => {
+                debug ("%" + int64.FORMAT + " bytes of %" + int64.FORMAT + " bytes copied.",
+                    current_num_bytes, release.size);
+            });
+        } catch (Error e) {
+            warning ("Could not download file: %s", e.message);
+            return;
+        }
+
+        if (!result) {
+            warning ("Download of %s was not succesfull", release.uri);
```

Comment:
These should probably be user-facing error dialogs right? Otherwise this would just fail silently